### ISSUE TITLE
BlendSearch: How do config constraints affect search performance?

### ISSUE THREAD

**Issue by bbudescu** [microsoft/FLAML#977]
As I can read in the [docs](https://microsoft.github.io/FLAML/docs/Use-Cases/Tune-User-Defined-Function/#config_constraints-vs-metric_constraints), when passing a valid argument for the `config_constraints` parameter of `flaml.tune.run`, if a particular hyperparameter configuration violates any of the constraints imposed, its evaluation is prevented so as to avoid wasting processing time, which is great.

However, when looking in the [BlendSearch code](https://github.com/microsoft/FLAML/blob/f404b66a4c20d421c9cee465a6796ca76a0a8992/flaml/tune/searcher/blendsearch.py#L856) I can see that it assigns an infinite loss to that configuration.

So, from a theoretical standpoint, given that the global search is based on a regressor which has a smoothness prior (I think you mentioned BlendSearch using a Tree-structured Parzen Estimator, which, as per my limited understanding of the [paper](https://papers.nips.cc/paper/4443-algorithms-for-hyper-parameter-optimization.pdf) uses truncated Gaussian mixture models), wouldn't that cause configuration that are topologically adjacent to the edge of the feasible region to also be estimated to have very low performance and, thus, decrease the probability for them to be sampled?

I mean, I am aware that robust regressors should have some capacity to model kinks in the surface they are estimating, but I wanted to ask how much of an extra cost would be incurred for the edge configurations to be taken into account, especially since it's possible that for some problems the optima have a high chance to lie right on the edge of feasibility.

**Comment by bbudescu** [comment]
This potential issue can be circumvented up to a certain degree using conditional sampling, but that's not very handy either. (#816, #988). E.g., I've been trying to enforce an order (inequality) constraint between two `randint`s, and it's not very easy without the potentially harmful `config_constraints` and without adding categorical `choice`s for each possible value.

**Comment by sonichi** [comment]
> As I can read in the [docs](https://microsoft.github.io/FLAML/docs/Use-Cases/Tune-User-Defined-Function/#config_constraints-vs-metric_constraints), when passing a valid argument for the `config_constraints` parameter of `flaml.tune.run`, if a particular hyperparameter configuration violates any of the constraints imposed, its evaluation is prevented so as to avoid wasting processing time, which is great.
> 
> However, when looking in the [BlendSearch code](https://github.com/microsoft/FLAML/blob/f404b66a4c20d421c9cee465a6796ca76a0a8992/flaml/tune/searcher/blendsearch.py#L856) I can see that it assigns an infinite loss to that configuration.
> 
> So, from a theoretical standpoint, given that the global search is based on a regressor which has a smoothness prior (I think you mentioned BlendSearch using a Tree-structured Parzen Estimator, which, as per my limited understanding of the [paper](https://papers.nips.cc/paper/4443-algorithms-for-hyper-parameter-optimization.pdf) uses truncated Gaussian mixture models), wouldn't that cause configuration that are topologically adjacent to the edge of the feasible region to also be estimated to have very low performance and, thus, decrease the probability for them to be sampled?
> 
> I mean, I am aware that robust regressors should have some capacity to model kinks (discontinuities) in the surface they are estimating, but I wanted to ask how much of an extra cost would be incurred for the edge configurations to be taken into account, especially since it's possible that for some problems the optima have a high chance to lie right on the edge of feasibility.

BlendSearch uses both model-based and model-free search. The model-free search won't be discouraged to sample near the edge. Have you observed any undesirable behavior when using `config_constraints` ?

**Comment by bbudescu** [comment]
Well, it's hard for me to measure on my problem where I don't know whether or not good solutions lie on the feasibility edge, and that's why I'm asking - because this kind of issue should be more apparent in a controlled environment, where the optima of the function to be optimized are known beforehand, and perhaps function evaluation is not as expensive.

In all fairness, the point you're making is correct - the local search might somewhat alleviate a potential shortcoming of the global search in this respect. However, I'm asking whether this hypothesis (i.e., that the local search is able to address this problem sufficiently) has been verified in practice, perhaps on a problem designed to highlight this capability (maybe within the research paper).

To provide some context:

I'm trying to sample from a space which looks, for our intents and purpose, something like "sample random integers `a` and `b` between 0 and 10 such that `a < b`". Now, I've tried several ways to do this, but to little avail:
- using `config_constraints` - I'm a bit concerned about performance (as explained in this issue), as I don't want to avoid, e.g., trying high `a`s or low `b`s independently altogether, but just in conjunction, and I'm not sure that the TPE is able to model this efficiently
- a conditional search space like `'b': flaml.tune.randint(1, 10). 'a': ray.tune.sample_from(lambda spec: flaml.tune.randint(0, b))` (this can be rewritten alternatively to sample `a` between 0 and 9 and `b` between `a` and 10); this crashes as explained in #988.
- sample `b` between 1 and 10, and then sample `a` as a float between 0 and 1 that should serve as a multiplier, i.e., we obtain the final value for `a` by multiplying the sampled float with `b`, and then rounding it to the nearest integer. However, this also has the drawback that multiple float values are mapped to the same integer, so you end up with many duplicate experiments. One way to alleviate this would be to store results for rounded values in `ray`'s  *object storage*, but that wouldn't remove the problem in its totality, because nothing prevents starting an identical trial (in parallel) before the original trial finishes.

So, if you, maybe have a different suggestion on how to do this properly, please give me an idea, as I'm all out :).

**Comment by bbudescu** [comment]
Perhaps I should try passing a custom pre-defined `ray.OptunaSearcher` instance, with the embedded space defined in terms of clean `ray.rune` APIs instead of using `flaml.tune`.

**Comment by sonichi** [comment]
> Well, it's hard for me to measure on my problem where I don't know whether or not good solutions lie on the feasibility edge, and that's why I'm asking - because this kind of issue should be more apparent in a controlled environment, where the optima of the function to be optimized are known beforehand, and perhaps function evaluation is not as expensive.
> 
> In all fairness, the point you're making is correct - the local search might somewhat alleviate a potential shortcoming of the global search in this respect. However, I'm asking whether this hypothesis (i.e., that the local search is able to address this problem sufficiently) has been verified in practice, perhaps on a problem designed to highlight this capability (maybe within the research paper).
> 
> To provide some context:
> 
> I'm trying to sample from a space which looks, for our intents and purpose, something like "sample random integers `a` and `b` between 0 and 10 such that `a < b`". Now, I've tried several ways to do this, but to little avail:
> 
> * using `config_constraints` - I'm a bit concerned about performance (as explained in this issue), as I don't want to avoid, e.g., trying high `a`s or low `b`s independently altogether, but just in conjunction, and I'm not sure that the TPE is able to model this efficiently
> * a conditional search space like `'b': flaml.tune.randint(1, 10), 'a': ray.tune.sample_from(lambda spec: flaml.tune.randint(0, b))` (this can be rewritten alternatively to sample `a` between 0 and 9 and `b` between `a` and 10); this crashes as explained in [this](https://github.com/microsoft/FLAML/issues/988#issuecomment-1502986531) comment of [flaml tune support for ray.tune.sample_from #988](https://github.com/microsoft/FLAML/issues/988).
> * sample `b` between 1 and 10, and then sample `a` as a float between 0 and 1 that should serve as a multiplier, i.e., we obtain the final value for `a` by multiplying the sampled float with `b`, and then rounding it to the nearest integer. However, this also has the drawback that multiple float values are mapped to the same integer, so you end up with many duplicate experiments. One way to alleviate this would be to store results for rounded values in `ray`'s  _object storage_ (i.e., using `ray.put` and `ray.get`), but that wouldn't remove the problem in its totality, because nothing prevents starting an identical trial (in parallel) before the original trial finishes.
> 
> So, if you, maybe have a different suggestion on how to do this properly, please give me an idea, as I'm all out :).

`config_constraints` is indeed used in the experiments in this paper: [FLAML: A Fast and Lightweight AutoML Library](https://www.microsoft.com/en-us/research/publication/flaml-a-fast-and-lightweight-automl-library/). CFO is used as the search method, not TPE.
